### PR TITLE
STL frontend: "Main" entry point

### PR DESCRIPTION
**Pull request by MatWise**
Expands the current entry point selection by an additional check for a block with the name "Main". Note: This does not reflect the typical behaviour of the Siemens TIA portal as the main block can have any name. In order to find the 'true' main block it is necessary to parse block properties, which is not possible yet. The check for "Main" is intended to be the last resort if no entry point is specified (or if the true main block could not be resolved in the future).

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c8c0f748383dbc11779d37da7cbed9466e6f7755).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119469000

File: src/statement-list/statement_list_entry_point.cpp [hunk 1]
```diff
@@ -30,6 +30,46 @@ Author: Matthias Weiss, matthias.weiss@diffblue.com
...
+/// \param symbol_table: Symbol table to search through.
+/// \param message_handler: Used for printing error messages.
+/// \param main_symbol_name: Name of the symbol to look for.
+/// \return False if there is exactly one match, true otherwise.
+static bool search_main_symbol(
+  const symbol_tablet &symbol_table,
+  message_handlert &message_handler,
+  const irep_idt &main_symbol_name)
+{
+  unsigned matches = 0u;
```
Comment (pkesseli) [inline src/statement-list/statement_list_entry_point.cpp]: Replace by boolean `found`.

File: src/statement-list/statement_list_entry_point.cpp [hunk 2]
```diff
@@ -30,6 +30,46 @@ Author: Matthias Weiss, matthias.weiss@diffblue.com
...
 
+/// Searches for symbols with the given name (which is considered to be the
+/// name of the main symbol) and returns false if there is exactly one symbol
+/// with that name in the given symbol table. Prints an error message and
+/// returns true if there are multiple or no matches.
+/// \param symbol_table: Symbol table to search through.
+/// \param message_handler: Used for printing error messages.
+/// \param main_symbol_name: Name of the symbol to look for.
+/// \return False if there is exactly one match, true otherwise.
+static bool search_main_symbol(
```
Comment (pkesseli) [inline src/statement-list/statement_list_entry_point.cpp]: Rename: `is_main_symbol_invalid`

File: src/statement-list/statement_list_entry_point.cpp [hunk 3]
```diff
@@ -162,46 +202,28 @@ bool statement_list_entry_point(
...
-    // --show-symbol-table). The missing entry symbol will be caught later.
-    return false;
+    if(search_main_symbol(
+         symbol_table, message_handler, ID_statement_list_main_function))
+      return true;
+    main_symbol_name = ID_statement_list_main_function;
   }
 
-  const symbolt &main = symbol_table.lookup_ref(main_symbol_name);
+  const symbolt &main_symbol = symbol_table.lookup_ref(main_symbol_name);
```
Comment (pkesseli) [inline src/statement-list/statement_list_entry_point.cpp]: Don't repeat type name in the variable name (unless there is a conflict): `main`

**Review by pkesseli — CHANGES_REQUESTED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: dee10712ba97df43b08aab5ccce9f88c787c7bb5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119640442

**Review by pkesseli — APPROVED** [review]

**Pull request merged**